mlflow/server/auth/__init__.py

```diff
@@ -1443,6 +1445,191 @@ def delete_scorer_permission():
+    )
+
+
+def _graphql_get_permission_for_run(run_id: str, username: str) -> Permission:
+    run = _get_tracking_store().get_run(run_id)
+    experiment_id = run.info.experiment_id
+    return _get_permission_from_store_or_default(
+        lambda: store.get_experiment_permission(experiment_id, username).permission
+    )
+
+
+def _graphql_get_permission_for_model(model_name: str, username: str) -> Permission:
+    return _get_permission_from_store_or_default(
+        lambda: store.get_registered_model_permission(model_name, username).permission
+    )
+
+
+def _graphql_can_read_experiment(experiment_id: str, username: str) -> bool:
+    return _graphql_get_permission_for_experiment(experiment_id, username).can_read
+
+
+def _graphql_can_read_run(run_id: str, username: str) -> bool:
+    return _graphql_get_permission_for_run(run_id, username).can_read
+
+
+def _graphql_can_read_model(model_name: str, username: str) -> bool:
+    return _graphql_get_permission_for_model(model_name, username).can_read
+
+
+class GraphQLAuthorizationMiddleware:
+    """
+    Graphene middleware that enforces per-object authorization for GraphQL queries.
+
+    This middleware checks user permissions before resolving protected fields.
+    It integrates with MLflow's basic-auth permission system.
+    """
+
+    PROTECTED_FIELDS = {
+        "mlflowGetExperiment",
+        "mlflowGetRun",
+        "mlflowListArtifacts",
+        "mlflowGetMetricHistoryBulkInterval",
+        "mlflowSearchRuns",
+        "mlflowSearchModelVersions",
+        "mlflowSearchDatasets",
+    }
+
+    def resolve(self, next, root, info, **args):
+        """
+        Middleware resolve function called for every field resolution.
+
+        Args:
+            next: The next resolver in the chain
+            root: The root value object
+            info: GraphQL resolve info containing field name and context
+            args: Field arguments as keyword arguments
+
+        Returns:
+            The resolved value or an error response
+        """
+        field_name = info.field_name
+
+        if field_name not in self.PROTECTED_FIELDS:
+            return next(root, info, **args)
+
+        try:
+            authorization = authenticate_request()
+            if isinstance(authorization, Response):
+                return None
+            username = authorization.username
+
+            if store.get_user(username).is_admin:
+                return next(root, info, **args)
+        except Exception:
+            _logger.warning("GraphQL authorization failed: auth system error", exc_info=True)
+            return None
+
+        try:
+            if not self._check_authorization(field_name, args, username):
+                _logger.debug(f"GraphQL authorization denied for {field_name} by user {username}")
+                return None
+        except MlflowException as e:
+            if e.error_code == ErrorCode.Name(RESOURCE_DOES_NOT_EXIST):
+                return None
```

Comment:
We don't need this because `_get_permission_from_store_or_default` already captures it